bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsHandler.java

```diff
@@ -0,0 +1,684 @@
+                                    "Writing to device failed.");
+                        }
+                    }
+                });
+            }
+        }
+    }
+
+    /**
+     * Writes a variable value to the Helios device
+     *
+     * @param variableName The variable name
+     * @param value The new value
+     * @return The value if the transaction succeeded, <tt>null</tt> otherwise
+     * @throws HeliosException Thrown if the variable is read-only or the provided value is out of range
+     */
+    public void writeValue(String variableName, String value) throws HeliosException {
+        if (this.vMap == null) {
+            this.handleError("Variable definition is unavailable.", ThingStatusDetail.CONFIGURATION_ERROR);
+            return;
+        } else {
+            HeliosVariable v = this.vMap.get(variableName);
+            if (!v.hasWriteAccess()) {
+                throw new HeliosException("Variable " + variableName + " is read-only");
+            } else if (!v.isInAllowedRange(value)) {
+                throw new HeliosException(
+                        "Value " + value + " is outside of allowed range of variable " + variableName);
+            } else if (this.comms != null) {
+                // write to device
+                String payload = v.getVariableString() + "=" + value;
+                final Semaphore lock = transactionLocks.get(this.comms.getEndpoint());
+                try {
+                    lock.acquire();
+                    this.comms.submitOneTimeWrite(
+                            new ModbusWriteRegisterRequestBlueprint(HeliosEasyControlsBindingConstants.UNIT_ID,
+                                    HeliosEasyControlsBindingConstants.START_ADDRESS,
+                                    new ModbusRegisterArray(preparePayload(payload)), true,
+                                    HeliosEasyControlsBindingConstants.MAX_TRIES),
+                            result -> {
+                                lock.release();
+                                updateStatus(ThingStatus.ONLINE);
+                            }, failureInfo -> {
+                                String errorMsg = failureInfo.getCause().getMessage();
+                                logger.warn("{} encountered error writing to device: {}",
+                                        HeliosEasyControlsHandler.class.getSimpleName(), errorMsg);
+                                lock.release();
+                                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, errorMsg);
+                            });
+                    // ensure the openHAB item is updated with the device's actual value
+                    scheduler.schedule(new Runnable() {
+                        @Override
+                        public void run() {
+                            readValue(variableName);
+                        }
+                    }, 1, TimeUnit.SECONDS);

```

Comment:
To avoid unnecessary updates I've attached this to the actual write command now.
Because without this the state update would take quite a while.

Reply:
Actually the Channel is instantly updated when the user sent a command to it. This is done by the framework. An update is only necessary if changing a Channel would affect another Channel's value.

Reply:
Then, fascinatingly, I've managed to implement a binding that counters this mechanism beautifully...
Chaning a value via a write request works fine and I see the effect on the device, but in OH the change will only become visibly when the next scheduled refresh is executed.
Do you have any idea what might be a possible cause?

Reply:
I observed this behavior, too, sometimes. It seems Basic UI misses to update the sitemap. Do you see the update in the console? Here is the explaining documentation: https://www.openhab.org/docs/developer/bindings/thing-xml.html#auto-update-policies

Reply:
I've just seen that I didn't actively update the channel state after writing to the device, just when reading from the device. I've now changed this to also update the channel state upon writing. I hope this is now more in line with the standard behaviour.

Reply:
Actually you only need to update the Channel when the change is not originated by the user. I just tested to change a Channel, which is not updated by the corresponding binding and I see its status changed on the console (`ItemStateChangedEvent`) when I send a command to it.

Reply:
Strange - if I don't update the channel the change is not done. I've tested this with a simple rule:
```
logInfo("TEST", triggeringItem + " - " + receivedCommand)
Thread::sleep(2000)
logInfo("TEST", triggeringItem.state.toString)
```
When it fires I see the correct command which triggered the rule but the item's state doesn't change. The second output is still the old state. However the command does physically change the device's state...

Reply:
When looking to the schedule, it might be a good idea to choose the approach which works for you at the moment to get this merged before 2.5.9. And then, file another PR when the above is clarified.

Reply:
The latest commit works fine for me and I'm happy to leave this as it is. If we find a more elegant approach later I'm happy to make the necessary changes.